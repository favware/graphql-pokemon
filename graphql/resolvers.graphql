type Query {
  "Gets the details on a Pokémon ability, using the ability name"
  getAbility("The ability to look up" ability: AbilitiesEnum!): Ability!

  """
  Gets details on a Pokémon ability, using a fuzzy search on name

  This can be used to find multiple results based on the query

  You can provide `take` to limit the amount of abilities to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.
  """
  getFuzzyAbility(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The ability to fuzzily search"
    ability: String!
  ): [Ability!]!

  """
  Gets details on a single Pokémon based on National Pokédex number

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemonByDexNumber(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The International PokéDex number of the Pokémon to look up"
    number: Int!
  ): Pokemon!

  """
  Gets details on a single Pokémon based on species name

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemon(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The Pokémon to look up"
    pokemon: PokemonEnum!
  ): Pokemon!

  """
  Gets details on one or more Pokémon based on species name

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getFuzzyPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The Pokémon to fuzzily search"
    pokemon: String!

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  """
  Returns a list of all the known Pokémon.

  For every Pokémon all the data on each requested field is returned.

  **_NOTE:_ To skip all CAP Pokémon, PokéStar Pokémon, and Missingno provide an `offset` of 87**

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  While the API will currently not rate limit the usage of this query, it may do so in the future.

  It is advisable to cache responses of this query.
  """
  getAllPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1273

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  "Gets the details on a Pokémon item, using the item name"
  getItem("The item to look up" item: ItemsEnum!): Item!

  """
  Gets details on a Pokémon item, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, `offset`, and `reverse` to modify this behaviour.
  """
  getFuzzyItem(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The item to fuzzily search"
    item: String!
  ): [Item!]!

  """
  Gets the learnsets for a given Pokémon and move.

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [MovesEnum!]!

    "The Pokémon for which to get the learnset"
    pokemon: PokemonEnum!
  ): Learnset!

  """
  Gets the learnset for a given Pokémon and move.

  A fuzzy search is performed to find a matching Pokémon and move

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getFuzzyLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [String!]!

    "The Pokémon for which to get the learnset"
    pokemon: String!
  ): Learnset!

  "Gets the details on a Pokémon move, using the move name"
  getMove("The move to look up" move: MovesEnum!): Move!

  """
  Gets details on a Pokémon move, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, "offset", and "reverse" to modify this behaviour.
  """
  getFuzzyMove(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The move to fuzzily search"
    move: String!
  ): [Move!]!

  "Gets the type matchup data for the given type or types"
  getTypeMatchup("The primary type to check" primaryType: TypesEnum!, "The secondary type to check" secondaryType: TypesEnum): TypeMatchup!
}
