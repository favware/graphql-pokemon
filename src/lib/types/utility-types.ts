/**
 * Removes `null`, `undefined` and `?` from an object
 * This essentially combines `Required<T>`,
 * `NonNullable<T>` and `NonUndefined<T>`
 */
export type NonNullish<T> = {
  [P in keyof T]-?: Exclude<T[P], undefined | null>;
};

export type UnwrapArray<T extends unknown[]> = T extends (infer U)[] ? U : never;

export type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
