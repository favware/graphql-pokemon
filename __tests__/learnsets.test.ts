import {formatResponse, gCall} from './testUtils/testUtils';
import {DataResponse} from './testUtils/types';
import {GraphQLError} from 'graphql';

const lvlmovesFragment = 'fragment lvlmoves on LearnsetLevelUpMove { name generation level }';
const movesFragment = 'fragment moves on LearnsetMove { name generation }';
const entryFragment = `
    fragment entry on LearnsetEntry {
      levelUpMoves { ...lvlmoves }
      virtualTransferMoves { ...moves }
      tutorMoves { ...moves }
      tmMoves { ...moves }
      eggMoves { ...moves }
      eventMoves { ...moves }
      dreamworldMoves { ...moves }
    }
  `;

describe('getPokemonLearnset', () => {
  const getPokemonLearnset = `
  ${lvlmovesFragment}
  ${movesFragment}
  ${entryFragment}

  query ($pokemon: Pokemon! $moves: [Moves!]! $generation: Int) {
    getPokemonLearnset(pokemon: $pokemon moves: $moves generation: $generation) {
      ...entry
    }
  }`;

  test('GIVEN a valid pokemon and move array THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: [ 'dragondance' ]},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.eventMoves).toContainEqual({name: 'dragondance', generation: 6});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a valid pokemon and single move THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: 'dragondance'},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.eventMoves).toContainEqual({name: 'dragondance', generation: 6});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a valid pokemon, move array and generation THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: [ 'dragondance' ], generation: 7},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.eventMoves).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a valid pokemon, a single move and generation THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: 'dragondance', generation: 7},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.eventMoves).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a valid pokemon and multiple moves THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: [ 'dragondance', 'dragonclaw' ]},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.tmMoves).toContainEqual({name: 'dragonclaw', generation: 6});
    expect(data.getPokemonLearnset.eventMoves).toContainEqual({name: 'dragondance', generation: 5});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a valid pokemon, multiple moves and generation THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: [ 'dragondance', 'dragonclaw' ], generation: 7},
    }).then(formatResponse) as DataResponse<'getPokemonLearnset'>;

    expect(data.getPokemonLearnset.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnset.tmMoves).toContainEqual({name: 'dragonclaw', generation: 7});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid pokemon THEN returns error', async () => {
    const data = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'totally_invalid_pokemon'},
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move THEN returns error', async () => {
    const data = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: 'totally_invalid_move'},
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid generation THEN returns error', async () => {
    const data = await gCall({
      source: getPokemonLearnset,
      variableValues: {pokemon: 'dragonite', moves: 'dragondance', generation: 0},
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});

describe('getPokemonLearnsetByFuzzy', () => {
  const getPokemonLearnsetByFuzzy = `
  ${lvlmovesFragment}
  ${movesFragment}
  ${entryFragment}

  query ($pokemon: String! $moves: [String!]! $generation: Int) {
    getPokemonLearnsetByFuzzy(pokemon: $pokemon moves: $moves generation: $generation) {
      ...entry
    }
  }`;

  test('GIVEN a fuzzy pokemon and exact move array THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnsetByFuzzy,
      variableValues: {pokemon: 'dragoni', moves: [ 'dragondance' ]},
    }).then(formatResponse) as DataResponse<'getPokemonLearnsetByFuzzy'>;

    expect(data.getPokemonLearnsetByFuzzy.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnsetByFuzzy.eventMoves).toContainEqual({name: 'dragondance', generation: 6});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an exact pokemon and fuzzy move array THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnsetByFuzzy,
      variableValues: {pokemon: 'dragonite', moves: [ 'dragondan' ]},
    }).then(formatResponse) as DataResponse<'getPokemonLearnsetByFuzzy'>;

    expect(data.getPokemonLearnsetByFuzzy.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnsetByFuzzy.eventMoves).toContainEqual({name: 'dragondance', generation: 6});
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy pokemon and fuzzy move array THEN returns LearnsetEntry', async () => {
    const {data} = await gCall({
      source: getPokemonLearnsetByFuzzy,
      variableValues: {pokemon: 'dragoni', moves: [ 'dragondan' ]},
    }).then(formatResponse) as DataResponse<'getPokemonLearnsetByFuzzy'>;

    expect(data.getPokemonLearnsetByFuzzy.levelUpMoves).toContainEqual({name: 'dragondance', generation: 7, level: 61});
    expect(data.getPokemonLearnsetByFuzzy.eventMoves).toContainEqual({name: 'dragondance', generation: 6});
    expect(data).toMatchSnapshot();
  });
});