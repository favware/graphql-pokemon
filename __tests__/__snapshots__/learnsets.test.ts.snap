// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPokemonLearnset GIVEN a valid pokemon and move array THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN a valid pokemon and multiple moves THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
      Object {
        "generation": 8,
        "name": "dragonclaw",
      },
      Object {
        "generation": 7,
        "name": "dragonclaw",
      },
      Object {
        "generation": 6,
        "name": "dragonclaw",
      },
      Object {
        "generation": 5,
        "name": "dragonclaw",
      },
      Object {
        "generation": 4,
        "name": "dragonclaw",
      },
      Object {
        "generation": 3,
        "name": "dragonclaw",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN a valid pokemon and single move THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN a valid pokemon, a single move and generation THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [],
    "levelUpMoves": Array [
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN a valid pokemon, move array and generation THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [],
    "levelUpMoves": Array [
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN a valid pokemon, multiple moves and generation THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnset": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [],
    "levelUpMoves": Array [
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 7,
        "name": "dragonclaw",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnset GIVEN an invalid generation THEN returns error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
}
`;

exports[`getPokemonLearnset GIVEN an invalid move THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$moves" got invalid value "totally_invalid_move"; Value "totally_invalid_move" does not exist in "Moves" enum.],
  ],
}
`;

exports[`getPokemonLearnset GIVEN an invalid pokemon THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$pokemon" got invalid value "totally_invalid_pokemon"; Value "totally_invalid_pokemon" does not exist in "Pokemon" enum.],
    [GraphQLError: Variable "$moves" of required type "[Moves!]!" was not provided.],
  ],
}
`;

exports[`getPokemonLearnsetByFuzzy GIVEN a fuzzy pokemon and exact move array THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnsetByFuzzy": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnsetByFuzzy GIVEN a fuzzy pokemon and fuzzy move array THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnsetByFuzzy": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;

exports[`getPokemonLearnsetByFuzzy GIVEN an exact pokemon and fuzzy move array THEN returns LearnsetEntry 1`] = `
Object {
  "getPokemonLearnsetByFuzzy": Object {
    "dreamworldMoves": Array [],
    "eggMoves": Array [],
    "eventMoves": Array [
      Object {
        "generation": 6,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "name": "dragondance",
      },
      Object {
        "generation": 3,
        "name": "dragondance",
      },
    ],
    "levelUpMoves": Array [
      Object {
        "generation": 8,
        "level": 62,
        "name": "dragondance",
      },
      Object {
        "generation": 7,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 6,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 5,
        "level": 61,
        "name": "dragondance",
      },
      Object {
        "generation": 4,
        "level": 53,
        "name": "dragondance",
      },
    ],
    "tmMoves": Array [
      Object {
        "generation": 8,
        "name": "dragondance",
      },
    ],
    "tutorMoves": Array [],
    "virtualTransferMoves": Array [],
  },
}
`;
